### ISSUE TITLE
Make getindex(::SubDataFrame, ::Union{Integer, Symbol}) return a view of column vector

### ISSUE THREAD

**Issue by GlenHertz** [JuliaData/DataFrames.jl#142]
When cleaning up data or doing signal processing, element by element, the user may need to modify individual values of a dataframe (df), subdataframes (sdf), and datavec (dv).  So that assignments can be made like:

``` Julia
sdf["B"][n] = (sdf["B"][n-1] + sdf["B"][n+1])/2 + sdf["C"][n-1]

#For Arrays and SubArrays:
a[:,1] #is a vector reference
a[2,3] #is a element reference
a[:,3][2] #is an element copy

#For DataFrames
df["A"] #is a dv reference
df[1,2] #is a element reference

#For SubDataFrames
sdf["A"] #is a dv copy
sdf[1,2] #is not valid

#For DataVecs
dv[3] #is an element reference
```

Let me know if I made any errors.

I would propose making `sdf["A"]` return a `SubDataVec` reference.  For consistency I think `sdf[1,2]` should be valid but since df's use named columns I don't see this being that popular.

Glen

**Comment by quinnj** [comment]
I don't think this is particularly relevant anymore, `view` is pretty standard and it's clear that it returns a "view" instead of copying, while all other indexing generally match Base array indexing behavior. The exception, I guess, is `df[3]`, which just returns the underlying column itself, but I think that's fair given the columnar-representation of DataFrames.

**Comment by nalimilan** [comment]
I think there's still something to improve: `sdf[1]` makes a copy of the first column with only the relevant indices. It would make more sense to return a view, since users would create a `SubDataFrame` precisely to avoid making copies.

**Comment by nalimilan** [comment]
@bkamins This is relevant for your `getindex` consistency improvements.

**Comment by bkamins** [comment]
I was not aware about this issue, but I thought the current functionality was intended.

Actually I think that there is a problem with `SubDataFrame` and `view` methods not `getindex`. We should reimplement `SubDataFrame` and `view` so that they take two arguments in general like `getindex` and passing one argument selects a column not a row.

Here is the example of the issue:

```
julia> sdf = view(DataFrame(x = 1:10, y = 1.0:10.0), 1:5)
5×2 SubDataFrame{UnitRange{Int64}}
│ Row │ x     │ y       │
│     │ Int64 │ Float64 │
├─────┼───────┼─────────┤
│ 1   │ 1     │ 1.0     │
│ 2   │ 2     │ 2.0     │
│ 3   │ 3     │ 3.0     │
│ 4   │ 4     │ 4.0     │
│ 5   │ 5     │ 5.0     │

julia> sdf[1]
5-element Array{Int64,1}:
 1
 2
 3
 4
 5

julia> @view sdf[1]
1×2 SubDataFrame{Array{Int64,1}}
│ Row │ x     │ y       │
│     │ Int64 │ Float64 │
├─────┼───────┼─────────┤
│ 1   │ 1     │ 1.0     │
```

The same happens when `sdf` would be a `DataFrame`.

In general I think that `sdf[1]` should make a copy as is currently implemented, but `view(sdf, 1)` should produce a view vector.

We could also add `DataFrame` constructor for `SubDataFrame`.

Any opinion? I can make a PR based on what we decide (it will in general be significantly breaking for `SubDataFrame` but otherwise we will not have a transparent consistency between `DataFrame` and `SubDataFrame`).

**Comment by bkamins** [comment]
Good point. `df[1]` and `df[:,1]` should create a view.

**Comment by bkamins** [comment]
I will push PRs fixing both things (I just have to give it some thought not to introduce other problems in corner cases).

**Comment by nalimilan** [comment]
> Good point. `df[1]` and `df[:,1]` should create a view.

Didn't you say that `df[:, 1]` should make a copy even for `DataFrame`?

**Comment by bkamins** [comment]
Well - you remember better than me; thank you. It should - I have checked this on current release and got confused (actually I thought it should be a copy and was surprised that it was not and that is why I have mentioned it 😄).

What I mean is that `DataFrame` and `SubDataFrame` should be consistent here. I will push the PR for `SubDataFrames` and a separate PR for documentation explaining current and target behavior of `getindex` for all objects in DataFrames.jl so that we and users have a clear understanding what to expect.

**Comment by bkamins** [comment]
Fixed in #1603
